vicky/src/lib/database/entities/task.rs

```diff
@@ -0,0 +1,415 @@
+                "FINISHED::ERROR" => TaskStatus::FINISHED(TaskResult::ERROR),
+                _ => TaskStatus::NEW,
+            }
+        }
+    }
+
+    impl From<&Task> for DbTask {
+        fn from(task: &Task) -> DbTask {
+            DbTask {
+                id: task.id,
+                display_name: task.display_name.clone(),
+                status: task.status.to_string(),
+                features: task.features.join("||"),
+                flake_ref_uri: task.flake_ref.flake.clone(),
+                flake_ref_args: task.flake_ref.args.join("||"),
+            }
+        }
+    }
+
+    #[derive(Selectable, Identifiable, Queryable, Debug)]
+    #[diesel(table_name = locks)]
+    struct DbLock {
+        id: i32,
+        task_id: Uuid,
+        name: String,
+        type_: String,
+    }
+
+    #[derive(Insertable, Debug)]
+    #[diesel(table_name = locks)]
+    struct NewDbLock {
+        task_id: Uuid,
+        name: String,
+        type_: String,
+    }
+
+    impl From<DbLock> for NewDbLock {
+        fn from(value: DbLock) -> Self {
+            NewDbLock {
+                task_id: value.task_id,
+                name: value.name,
+                type_: value.type_,
+            }
+        }
+    }
+
+    impl DbLock {
+        fn from_lock(lock: &Lock, task_id: Uuid) -> Self {
+            match lock {
+                Lock::WRITE { name } => DbLock {
+                    id: -1,
+                    task_id,
+                    name: name.clone(),
+                    type_: "WRITE".to_string(),
+                },
+                Lock::READ { name } => DbLock {
+                    id: -1,
+                    task_id,
+                    name: name.clone(),
+                    type_: "READ".to_string(),
+                },
+            }
+        }
+    }
+
+    impl From<DbLock> for Lock {
+        fn from(lock: DbLock) -> Lock {
+            match lock.type_.as_str() {
+                "WRITE" => Lock::WRITE { name: lock.name },
+                "READ" => Lock::READ { name: lock.name },
+                _ => panic!(
+                    "Can't parse lock from database lock. Database corrupted? \
+                Expected READ or WRITE but found {} as type at key {}.",
+                    lock.type_, lock.id
+                ),
+            }
+        }
+    }
+
+    #[database("postgres_db")]
+    pub struct Database(diesel::PgConnection);
+
+    pub trait TaskDatabase {
+        fn get_all_tasks(&mut self) -> Result<Vec<Task>, VickyError>;
+        fn get_task(&mut self, task_id: Uuid) -> Result<Option<Task>, VickyError>;
+        fn put_task(&mut self, task: &Task) -> Result<(), VickyError>;
+        fn update_task(&mut self, task: &Task) -> Result<(), VickyError>;
+    }
+
+    impl TaskDatabase for diesel::pg::PgConnection {
+        fn get_all_tasks(&mut self) -> Result<Vec<Task>, VickyError> {
+            // very evil >>:(
+            use self::locks::dsl::*;
```

Comment:
I am not a huge fan of those imports, but I did not find a better way, yet.

Reply:
The """fix""" was to put them in the functions where they are needed. I don't think there's a better way to do this because you rely on these names in all the diesel statements. Like, task_id.eq()....